runtime/src/main/java/org/corfudb/runtime/view/AddressSpaceView.java

```diff
@@ -417,68 +425,72 @@ public void invalidateServerCaches() {
         });
     }
 
-    /** Force the client cache to be invalidated. */
+    /**
+     * Force the client cache to be invalidated.
+     */
     public void invalidateClientCache() {
         readCache.invalidateAll();
     }
 
     /**
      * Fetch a collection of addresses for insertion into the cache.
      *
-     * @param addresses collection of addresses to read from.
-     * @return A result to be cached
+     * @param addresses    collection of addresses to read from.
+     * @param waitForWrite flag whether wait for write is required or hole fill directly.
+     * @return a map of read addresses.
      */
-    public @Nonnull
-    Map<Long, ILogData> fetchAll(Iterable<Long> addresses) {
+    @Nonnull
+    public Map<Long, ILogData> fetchAll(Iterable<Long> addresses, boolean waitForWrite) {
         Map<Long, ILogData> result = new HashMap<>();
 
         Iterable<List<Long>> batches = Iterables.partition(addresses,
-            runtime.getParameters().getBulkReadSize());
+                runtime.getParameters().getBulkReadSize());
 
         for (List<Long> batch : batches) {
             try {
-                //doesn't handle the case where some address have a different replication mode
+                // Doesn't handle the case where some address have a different replication mode
                 result.putAll(layoutHelper(e -> e.getLayout()
                         .getReplicationMode(batch.iterator().next())
                         .getReplicationProtocol(runtime)
-                        .readAll(e, batch)));
+                        .readAll(e, batch, waitForWrite)));
             } catch (Exception e) {
                 log.error("cacheFetch: Couldn't read addresses {}", batch, e);
                 throw new UnrecoverableCorfuError(
-                    "Unexpected error during cacheFetch", e);
+                        "Unexpected error during cacheFetch", e);
             }
         }
 
-        for (Long address : result.keySet()) {
-            checkLogData(address, result.get(address));
-        }
+        result.forEach(this::checkLogData);
+
         return result;
     }
 
     /**
-     * Fetch a range of addresses.
+     * Fetch a closed range of addresses. Batching should be handled by caller.
+     * Note: this method does not leverage caching ability.
      *
-     * @param addresses collection of addresses to read from.
-     * @return A result to be cached
+     * @param addresses closed range of addresses to read from.
+     * @return a map of read addresses.
      */
-    public @Nonnull
-    Map<Long, ILogData> fetchRange(Set<Long> addresses) {
-        Map<Long, ILogData> result = layoutHelper(e -> e.getLayout().getReplicationMode(addresses.iterator().next())
+    @Nonnull
+    public Map<Long, ILogData> fetchRange(Range<Long> addresses, boolean waitForWrite) {
+        Map<Long, ILogData> result = layoutHelper(e -> e.getLayout()
+                .getReplicationMode(addresses.lowerEndpoint())
```

Comment:
Btw this can be incorrect if the read spans two segments that don't have the same replication protocol.

Reply:
Yes this is an existing issue, and there is a comment say we don't handle this case. Will copy that comment here as well.

Reply:
Let's fix this issue if we are refactoring this code right now.
Issue: #1795

Reply:
#1795 Addressed.